README.md
# JMAP

A pure-Haskell implementation of the JMAP core protocol, as defined in
[RFC 8620](https://www.rfc-editor.org/rfc/rfc8620.html).


CHANGELOG.md
# Changelog for `jmap`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


test/JMAP/Types/BaseSpec.hs
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}

module JMAP.Types.BaseSpec (spec) where

import qualified Data.Aeson as A
import qualified Data.Aeson.Key as AK
import qualified Data.Aeson.Types as AT
import qualified Data.Text as T
import qualified Data.Vector as V
import           JMAP.Types.Arbitrary ()
import           JMAP.Types.Base
import           Test.Hspec
import           Test.Hspec.QuickCheck
import           Test.QuickCheck hiding (isSuccess)

spec :: Spec
spec = do
  describe "Id" $ do
    context "FromJSON" $ do
      prop "round trips" $ \(a :: Id)
        -> A.eitherDecode (A.encode a) === Right a
      it "rejects empty strings" $ do
        let r = A.fromJSON (A.String "") :: A.Result Id
        r `shouldNotSatisfy` isSuccess
      it "rejects any string longer than 255 characters" $ do
        let r = A.fromJSON (A.String $ T.replicate 256 "a") :: A.Result Id
        r `shouldNotSatisfy` isSuccess
      it "rejects characters outside base64url" $ do
        let r1 = A.fromJSON (A.String "=") :: A.Result Id
        r1 `shouldNotSatisfy` isSuccess
        let r2 = A.fromJSON (A.String ".com") :: A.Result Id
        r2 `shouldNotSatisfy` isSuccess
        let r3 = A.fromJSON (A.String "recherché") :: A.Result Id
        r3 `shouldNotSatisfy` isSuccess
        let r4 = A.fromJSON (A.String "a lot") :: A.Result Id
        r4 `shouldNotSatisfy` isSuccess
    context "FromJSONKey" $ do
      prop "round trips" $ \(a :: Id)
        -> let A.ToJSONKeyText f _ = A.toJSONKey
               A.FromJSONKeyTextParser g = A.fromJSONKey
           in AT.parse g (AK.toText $ f a) === A.Success a
      it "rejects empty strings" $ do
        let A.FromJSONKeyTextParser f = A.fromJSONKey
            r = AT.parse f "" :: A.Result Id
        r `shouldNotSatisfy` isSuccess
      it "rejects any string longer than 255 characters" $ do
        let A.FromJSONKeyTextParser f = A.fromJSONKey
            r = AT.parse f (T.replicate 256 "a") :: A.Result Id
        r `shouldNotSatisfy` isSuccess
      it "rejects characters outside base64url" $ do
        let A.FromJSONKeyTextParser f = A.fromJSONKey
            r1 = AT.parse f "base64=" :: A.Result Id
        r1 `shouldNotSatisfy` isSuccess
        let r2 = AT.parse f "ietf.org" :: A.Result Id
        r2 `shouldNotSatisfy` isSuccess
        let r3 = AT.parse f "Mëtal" :: A.Result Id
        r3 `shouldNotSatisfy` isSuccess
        let r4 = AT.parse f "S P A A A A C E" :: A.Result Id
        r4 `shouldNotSatisfy` isSuccess
    context "ToJSON" $ do
      it "encodes to a JSON String" $ do
        a <- generate (arbitrary :: Gen Id)
        A.toJSON a `shouldSatisfy` isString
    context "ToJSONKey" $ do
      it "encodes in a text-like way" $ do
        a <- generate (arbitrary :: Gen Id)
        let A.ToJSONKeyText f _ = A.toJSONKey
        AK.toText (f a) `shouldBe` fromId a

  describe "UInt" $ do
    context "Bounded" $ do
      it "has correct lower bound" $ do
        fromUInt minBound `shouldBe` (0 :: Integer)
      it "has correct upper bound" $ do
        fromUInt maxBound `shouldBe` (9007199254740989 :: Integer)
    context "FromJSON" $ do
      prop "round trips" $ \(a :: UInt)
        -> A.eitherDecode (A.encode a) === Right a
      it "rejects negative numbers" $ do
        let r = A.fromJSON (A.Number (-1)) :: A.Result UInt
        r `shouldNotSatisfy` isSuccess
      it "rejects numbers above the cap" $ do
        let r = A.fromJSON (A.Number 9007199254740990) :: A.Result UInt
        r `shouldNotSatisfy` isSuccess
    context "ToJSON" $ do
      it "encodes to a JSON Number" $ do
        a <- generate (arbitrary :: Gen UInt)
        A.toJSON a `shouldSatisfy` isNumber

  describe "Invocation" $ do
    context "FromJSON" $ do
      prop "round trips" $ \(a :: Invocation)
        -> A.eitherDecode (A.encode a) === Right a
    context "ToJSON" $ do
      it "encodes as a 3-element JSON Array" $ do
        a <- generate arbitrary
        let A.Array v = A.toJSON a
        V.length v `shouldBe` 3
        V.unsafeIndex v 0 `shouldBe` A.toJSON (name a)
        V.unsafeIndex v 1 `shouldBe` A.toJSON (arguments a)
        V.unsafeIndex v 2 `shouldBe` A.toJSON (method_call_id a)

-- Helpers

isSuccess :: A.Result a -> Bool
isSuccess (A.Error _) = False
isSuccess (A.Success _) = True

isNumber :: A.Value -> Bool
isNumber (A.Number _) = True
isNumber _ = False

isString :: A.Value -> Bool
isString (A.String _) = True
isString _ = False


test/JMAP/Types/Arbitrary.hs
{-# OPTIONS_GHC -Wno-orphans #-}

module JMAP.Types.Arbitrary () where

import qualified Data.Char as C
import qualified Data.Text as T
import           Data.Text.Arbitrary ()
import qualified Generic.Random as R
import           GHC.Data.Maybe (rightToMaybe)
import           JMAP.Types.Base (fromUInt, Id(..), Invocation, UInt(..))
import           JMAP.Types.Error (ErrorResponse(..), ErrorType)
import           JMAP.Types.Request (Request)
import           JMAP.Types.Response (Response)
import           JMAP.Types.Session (Account, Session)
import           Refined
import           Test.QuickCheck
import           Test.QuickCheck.Instances.UnorderedContainers ()

instance Arbitrary Account where
  arbitrary = R.genericArbitraryU
  shrink = genericShrink

instance Arbitrary ErrorResponse where
  arbitrary = ErrorResponse <$> R.genericArbitraryU
  shrink (ErrorResponse e) = ErrorResponse <$> genericShrink e

instance Arbitrary ErrorType where
  arbitrary = R.genericArbitraryU
  shrink = genericShrink

instance Arbitrary Id where
  arbitrary = Id <$> safeCharVec `suchThatMap` (maybeRefine . T.pack)
    where safeCharVec = chooseInt (1,255) >>= flip vectorOf safeChar
          safeChar = arbitraryASCIIChar `suchThat` isSafeChar
          isSafeChar c = c == '-' || c == '_' || C.isAlphaNum c
          maybeRefine = rightToMaybe . refine
  shrink = genericShrink

instance Arbitrary Invocation where
  arbitrary = R.genericArbitraryU
  shrink = genericShrink

instance Arbitrary Request where
  arbitrary = R.genericArbitraryU
  shrink = genericShrink

instance Arbitrary Response where
  arbitrary = R.genericArbitraryU
  shrink = genericShrink

instance Arbitrary Session where
  arbitrary = R.genericArbitraryU
  shrink = genericShrink

instance Arbitrary UInt where
  arbitrary = UInt <$> chooseInteger bs `suchThatMap` imRefine
    where bs = (fromUInt minBound, fromUInt maxBound)
          imRefine = rightToMaybe . refine . fromInteger
  shrink = genericShrink


test/JMAP/Types/ErrorSpec.hs
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}

module JMAP.Types.ErrorSpec (spec) where

import qualified Data.Aeson as A
import qualified Data.Aeson.KeyMap as AK
import           JMAP.Types.Arbitrary ()
import           JMAP.Types.Error
import           Refined (unrefine)
import           Test.Hspec
import           Test.Hspec.QuickCheck
import           Test.QuickCheck

spec :: Spec
spec = do
  describe "ErrorResponse" $ do
    context "FromJSON" $ do
      prop "round trips" $ \(a :: ErrorResponse)
        -> A.eitherDecode (A.encode a) === Right a

  describe "unknownCapability" $ do
    it "has all the common fields" $ do
      u <- generate arbitrary
      let A.Object o = A.toJSON $ unknownCapability u
      AK.lookup "type" o `shouldBe`
        Just (A.String "urn:ietf:params:jmap:error:unknownCapability")
      AK.lookup "title" o `shouldBe`
        Just (A.String "Requested capability is not known to the server")
      AK.lookup "detail" o `shouldBe`
        Just (A.String $ "The client included the capability '"<>u<>"' in the 'using' property of the request. The server does not support that capability.")
    it "has an unknownCapability field" $ do
      u <- generate arbitrary
      let A.Object o = A.toJSON $ unknownCapability u
      AK.lookup "unknownCapability" o `shouldBe` Just (A.String u)

  describe "notJSON" $ do
    it "has all the common fields" $ do
      let A.Object o = A.toJSON notJSON
      AK.lookup "type" o `shouldBe`
        Just (A.String "urn:ietf:params:jmap:error:notJSON")
      AK.lookup "title" o `shouldBe`
        Just (A.String "Server was unable to parse request as I-JSON")
      AK.lookup "detail" o `shouldBe`
        Just (A.String "The body of the request was either not a valid JSON document as defined in RFC-8259, or does not obey the conventions of the Internet JSON (I-JSON) format as defined in RFC-7493.")

  describe "notRequest" $ do
    it "has all the common fields" $ do
      let A.Object o = A.toJSON notRequest
      AK.lookup "type" o `shouldBe`
        Just (A.String "urn:ietf:params:jmap:error:notRequest")
      AK.lookup "title" o `shouldBe`
        Just (A.String "Server did not recognize JSON document as a Request object")
      AK.lookup "detail" o `shouldBe`
        Just (A.String "The server recognized the body of the request as valid I-JSON, but didn’t recognize that JSON document as a Request object as defined in the JMAP spec (RFC-8620 § 3.3).")

  describe "limit" $ do
    it "has all the common fields" $ do
      l <- generate arbitrary
      let A.Object o = A.toJSON $ limit l
      AK.lookup "type" o `shouldBe`
        Just (A.String "urn:ietf:params:jmap:error:limit")
      AK.lookup "title" o `shouldBe`
        Just (A.String "Server did not process request as it would exceed limit")
      AK.lookup "detail" o `shouldBe`
        Just (A.String $ "The request was not processed as it would have exceeded the '"<>unrefine l<>"' limit of the capability 'urn:ietf:params:jmap:core'. Capabilities and their limits may be found in the Session object returned from the server in accordance with the JMAP spec (RFC-8620 § 2).")
    it "has a limit field" $ do
      l <- generate arbitrary
      let A.Object o = A.toJSON $ limit l
      AK.lookup "limit" o `shouldBe` Just (A.String $ unrefine l)

{-
  Example errors from Section 3.6.1.1:

  {
    "type": "urn:ietf:params:jmap:error:unknownCapability",
    "status": 400,
    "detail": "The Request object used capability 'https://example.com/apis/foobar', which is not supported by this server."
  }

  {
    "type": "urn:ietf:params:jmap:error:limit",
    "limit": "maxSizeRequest",
    "status": 400,
    "detail": "The request is larger than the server is willing to process."
  }
-}


src/JMAP/Types/Base.hs
{-# OPTIONS_GHC -Wno-type-defaults #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeOperators #-}

module JMAP.Types.Base
  ( Collation
  , Id(..)
  , fromId
  , Invocation(..)
  , Limit
  , UInt(..)
  , fromUInt
  , URI
  ) where

import qualified Data.Aeson as A
import qualified Data.Aeson.Types as AT
import qualified Data.Char as C
import qualified Data.Hashable as H
import qualified Data.Maybe as U
import           Data.Text (Text)
import qualified Data.Text as T
import           Data.Typeable (typeRep)
import qualified Data.Vector as V
import           Data.Word (Word64)
import           GHC.Generics (Generic)
import           GHC.TypeLits
import           Refined

-- | Predicate used by @Id@.
data SafeChars = SafeChars
  deriving (Generic)

instance Predicate SafeChars Text where
  validate p x = if T.all isSafeChar x
                 then Nothing
                 else throwRefineOtherException
                      (typeRep p)
                      "Text contains unsafe characters [RFC 8620 § 1.2]."

isSafeChar :: Char -> Bool
isSafeChar c = c == '-' || c == '_' || (C.isAscii c && C.isAlphaNum c)


-- | The 'Id' type defined in Section 1.2. All record ids are assigned
-- by the server and are immutable.
--
-- Where "Id" is given as a data type, it means a "String" of at least
-- 1 and a maximum of 255 octets in size, and it MUST only contain
-- characters from the "URL and Filename Safe" base64 alphabet, as
-- defined in Section 5 of [RFC4648], excluding the pad character
-- ("="). This means the allowed characters are the ASCII alphanumeric
-- characters ("A-Za-z0-9"), hyphen ("-"), and underscore ("_").
--
-- These characters are safe to use in almost any context (e.g.,
-- filesystems, URIs, and IMAP atoms). For maximum safety, servers
-- SHOULD also follow defensive allocation strategies to avoid
-- creating risks where glob completion or data type detection may be
-- present (e.g., on filesystems or in spreadsheets). In particular,
-- it is wise to avoid:
--
-- * Ids starting with a dash
-- * Ids starting with digits
-- * Ids that contain only digits
-- * Ids that differ only by ASCII case (for example, A vs. a)
-- * the specific sequence of three characters "NIL" (because this
--   sequence can be confused with the IMAP protocol expression of the
--   null value)
newtype Id = Id (Refined (NonEmpty && SizeLessThan 256 && SafeChars) Text)
  deriving (Eq, A.FromJSONKey, Generic, Show, A.ToJSONKey)

instance A.FromJSON Id

instance H.Hashable Id

instance A.ToJSON Id where
  toEncoding = A.genericToEncoding A.defaultOptions

fromId :: Id -> Text
fromId (Id t) = unrefine t


-- TODO: Smuggle this into types
maxNum :: SomeNat
maxNum = U.fromJust $ someNatVal $ (2^53) - 1

-- | The "Unsigned Int" defined in Section 1.3. The value MUST be in
-- the range 0 <= value <= 2^53-1.
newtype UInt = UInt (Refined (To 9007199254740989) Word64)
  deriving (Eq, Generic, Show)

instance Bounded UInt where
  minBound = UInt $$(refineTH 0)
  maxBound = UInt $$(refineTH 9007199254740989)

instance A.FromJSON UInt

instance A.ToJSON UInt where
  toEncoding = A.genericToEncoding A.defaultOptions

fromUInt :: Num b => UInt -> b
fromUInt (UInt r) = fromIntegral $ unrefine r


-- | RFC 4790
type Collation = Text


-- | Placeholder for when I need to do some actual processing on URIs.
type URI = Text


-- | The Invocation data type as described in Section 3.2. Method
-- calls and responses are represented by the *Invocation* data
-- type. This is a tuple, represented as a JSON array containing three
-- elements:
data Invocation = Invocation
  { name :: !Text
    -- ^ 1. A "String" *name* of the method to call or of the
    -- response.
  , arguments :: !A.Value
    -- ^ 2. A "String[*]" object containing named *arguments* for that
    -- method or response.
  , method_call_id :: !Text
    -- ^ 3. A "String" *method call id*: an arbitrary string from the
    -- client to be echoed back with the responses emitted by that
    -- method call (a method may return 1 or more responses, as it may
    -- make implicit calls to other methods; all responses initiated
    -- by this method call get the same method call id in the
    -- response).
  } deriving (Eq, Generic, Show)

instance A.FromJSON Invocation where
  parseJSON = A.withArray "Invocation" $ \v -> Invocation
    <$> parseIndex v 0
    <*> parseIndex v 1
    <*> parseIndex v 2

-- | Helper: extract 'A.Value' at a given index.
parseIndex :: A.FromJSON a => A.Array -> Int -> AT.Parser a
parseIndex v i = case V.indexM v i of
  Nothing -> fail $ "index "<>show i<>" not found"
  Just a -> A.parseJSON a A.<?> AT.Index i

instance A.ToJSON Invocation where
  toJSON a = A.Array $ flip V.unfoldr (0 :: Natural) $ \i ->
    case i of
      0 -> Just (A.toJSON $ name a, 1)
      1 -> Just (A.toJSON $ arguments a, 2)
      2 -> Just (A.toJSON $ method_call_id a, 3)
      _ -> Nothing
  toEncoding a = A.foldable
    [ A.toJSON $ name a
    , A.toJSON $ arguments a
    , A.toJSON $ method_call_id a
    ]


-- | Placeholder for referencing a service limit in an
-- 'ErrorRepresentation'.
type Limit = Refined NonEmpty Text


src/JMAP/Types/Request.hs
{-# LANGUAGE DeriveGeneric #-}

module JMAP.Types.Request
  ( Request(..)
  ) where

import qualified Data.Aeson as A
import           Data.HashMap.Strict (HashMap)
import           GHC.Generics (Generic)
import           JMAP.Types.Base

-- FIXME: Spec demands not just JSON, but I-JSON

-- | The Request object as described in Section 3.3.
data Request = Request
  { using :: ![URI]
    -- ^ The set of capabilities the client wishes to use. The client
    -- MAY include capability identifiers even if the method calls it
    -- makes do not utilise those capabilities. The server advertises
    -- the set of specifications it supports in the Session object
    -- (see Section 2), as keys on the "capabilities" property.
  , methodCalls :: ![Invocation]
    -- ^ An array of method calls to process on the server. The method
    -- calls MUST be processed sequentially, in order.
  , createdIds :: !(Maybe (HashMap Id Id))
    -- ^ A map of a (client-specified) creation id to the id the
    -- server assigned when a record was successfully created.
    --
    -- As described later in this specification, some records may have
    -- a property that contains the id of another record. To allow
    -- more efficient network usage, you can set this property to
    -- reference a record created earlier in the same API
    -- request. Since the real id is unknown when the request is
    -- created, the client can instead specify the creation id it
    -- assigned, prefixed with a "#" (see Section 5.3 for more
    -- details).
    --
    -- As the server processes API requests, any time it successfully
    -- creates a new record, it adds the creation id to this map (see
    -- the "create" argument to /set in Section 5.3), with the
    -- server-assigned real id as the value. If it comes across a
    -- reference to a creation id in a create/update, it looks it up
    -- in the map and replaces the reference with the real id, if
    -- found.
    --
    -- The client can pass an initial value for this map as the
    -- "createdIds" property of the Request object. This may be an
    -- empty object. If given in the request, the response will also
    -- include a createdIds property. This allows proxy servers to
    -- easily split a JMAP request into multiple JMAP requests to send
    -- to different servers. For example, it could send the first two
    -- method calls to server A, then the third to server B, before
    -- sending the fourth to server A again. By passing the createdIds
    -- of the previous response to the next request, it can ensure all
    -- of these still resolve. See Section 5.8 for further discussion
    -- of proxy considerations.
  } deriving (Eq, Generic, Show)

instance A.FromJSON Request

instance A.ToJSON Request where
  toEncoding = A.genericToEncoding A.defaultOptions


src/JMAP/Types/Session.hs
{-# LANGUAGE DeriveGeneric #-}

module JMAP.Types.Session
  ( Session(..)
  , Capabilities
  , Accounts
  , Account(..)
  ) where

import qualified Data.Aeson as A
import           Data.HashMap.Strict (HashMap)
import           Data.Text (Text)
import           GHC.Generics (Generic)
import           JMAP.Types.Base

-- | The JMAP Session Resource of Section 2. A successful
-- authenticated GET request to the JMAP Session resource MUST return
-- a JSON-encoded Session object, giving details about the data and
-- capabilities the server can provide to the client given those
-- credentials.  It has the following properties:
data Session = Session
  { capabilities :: !Capabilities
    -- ^ An object specifying the capabilities of this server. Each
    -- key is a URI for a capability supported by the server. The
    -- value for each of these keys is an object with further
    -- information about the server's capabilities in relation to that
    -- capability. The capabilities object MUST include a property
    -- called "urn:ietf:params:jmap:core".
  , accounts :: !Accounts
    -- ^ A map of an account id to an Account object for each account
    -- (see Section 1.6.2) the user has access to.
  , primaryAccounts :: !(HashMap URI Id)
    -- ^ A map of capability URIs (as found in accountCapabilities) to
    -- the account id that is considered to be the user's main or
    -- default account for data pertaining to that capability. If no
    -- account being returned belongs to the user, or in any other way
    -- there is no appropriate way to determine a default account,
    -- there MAY be no entry for a particular URI, even though that
    -- capability is supported by the server (and in the capabilities
    -- object). "urn:ietf:params:jmap:core" SHOULD NOT be present.
  , username :: !(Maybe Text)
    -- ^ The username associated with the given credentials, or the empty string if none.
  , apiUrl :: !(Maybe Text)
    -- ^ The URL to use for JMAP API requests.
  , downloadUrl :: !(Maybe Text)
    -- ^ The URL endpoint to use when downloading files, in URI
    -- Template (level 1) format [RFC6570]. The URL MUST contain
    -- variables called "accountId", "blobId", "type", and "name". The
    -- use of these variables is described in Section 6.2. Due to
    -- potential encoding issues with slashes in content types, it is
    -- RECOMMENDED to put the "type" variable in the query section of
    -- the URL.
  , uploadUrl :: !(Maybe Text)
    -- ^ The URL endpoint to use when uploading files, in URI Template
    -- (level 1) format [RFC6570]. The URL MUST contain a variable
    -- called "accountId". The use of this variable is described in
    -- Section 6.1.
  , eventSourceUrl :: !(Maybe Text)
    -- ^ The URL to connect to for push events, as described in
    -- Section 7.3, in URI Template (level 1) format [RFC6570]. The
    -- URL MUST contain variables called "types", "closeafter", and
    -- "ping". The use of these variables is described in Section 7.3.
  , state :: !(Maybe Text)
    -- ^ A (preferably short) string representing the state of this
    -- object on the server. If the value of any other property on the
    -- Session object changes, this string will change. The current
    -- value is also returned on the API Response object (see Section
    -- 3.4), allowing clients to quickly determine if the session
    -- information has changed (e.g., an account has been added or
    -- removed), so they need to refetch the object.
  } deriving (Eq, Generic, Show)

instance A.FromJSON Session

instance A.ToJSON Session where
  toEncoding = A.genericToEncoding A.defaultOptions


-- | Section 2 gives the type as "String[Object]". Specifications for
-- future capabilities will define their own properties on the
-- capabilities object. Servers MAY advertise vendor-specific JMAP
-- extensions, as described in Section 1.8. To avoid conflict, an
-- identifier for a vendor-specific extension MUST be a URL with a
-- domain owned by the vendor. Clients MUST opt in to any capability
-- it wishes to use (see Section 3.3).
type Capabilities = HashMap URI A.Value


-- | A map of an account id to an Account object for each account (see
-- Section 1.6.2) the user has access to.
type Accounts = HashMap Id Account


data Account = Account
  { name :: !Text
    -- ^ A user-friendly string to show when presenting content from
    -- this account, e.g., the email address representing the owner of
    -- the account.
  , isPersonal :: !Bool
    -- ^ This is true if the account belongs to the authenticated user
    -- rather than a group account or a personal account of another
    -- user that has been shared with them.
  , isReadOnly :: !Bool
    -- ^ This is true if the entire account is read-only.
  , accountCapabilities :: !Capabilities
    -- ^ The set of capability URIs for the methods supported in this
    -- account. Each key is a URI for a capability that has methods
    -- you can use with this account. The value for each of these keys
    -- is an object with further information about the account's
    -- permissions and restrictions with respect to this capability,
    -- as defined in the capability's specification.
    --
    -- The client MUST ignore any properties it does not
    -- understand. The server advertises the full list of capabilities
    -- it supports in the capabilities object, as defined above. If
    -- the capability defines new methods, the server MUST include it
    -- in the accountCapabilities object if the user may use those
    -- methods with this account. It MUST NOT include it in the
    -- accountCapabilities object if the user cannot use those methods
    -- with this account.
    --
    -- For example, you may have access to your own account with mail,
    -- calendars, and contacts data and also a shared account that
    -- only has contacts data (a business address book, for
    -- example). In this case, the accountCapabilities property on the
    -- first account would include something like
    -- "urn:ietf:params:jmap:mail", "urn:ietf:params:jmap:calendars",
    -- and "urn:ietf:params:jmap:contacts", while the second account
    -- would just have the last of these.
    --
    -- Attempts to use the methods defined in a capability with one of
    -- the accounts that does not support that capability are rejected
    -- with an "accountNotSupportedByMethod" error (see "Method-Level
    -- Errors", Section 3.6.2).
  } deriving (Eq, Generic, Show)

instance A.FromJSON Account

instance A.ToJSON Account where
  toEncoding = A.genericToEncoding A.defaultOptions


src/JMAP/Types/Response.hs
{-# LANGUAGE DeriveGeneric #-}

module JMAP.Types.Response
  ( Response(..)
  ) where

import qualified Data.Aeson as A
import           Data.HashMap.Strict (HashMap)
import           Data.Text (Text)
import           GHC.Generics (Generic)
import           JMAP.Types.Base

-- | The Response object as described in Section 3.4.
data Response = Response
  { methodResponses :: ![Invocation]
    -- ^ An array of responses, in the same format as the
    -- "methodCalls" on the Request object. The output of the methods
    -- MUST be added to the "methodResponses" array in the same order
    -- that the methods are processed.
    --
    -- Unless otherwise specified, if the method call completed
    -- successfully, its response name is the same as the method name
    -- in the request.
  , createdIds :: !(Maybe (HashMap Id Id))
    -- ^ A map of a (client-specified) creation id to the id the
    -- server assigned when a record was successfully created. This
    -- MUST include all creation ids passed in the original createdIds
    -- parameter of the Request object, as well as any additional ones
    -- added for newly created records.
  , sessionState :: !Text
    -- ^ The current value of the "state" string on the Session
    -- object, as described in Section 2. Clients may use this to
    -- detect if this object has changed and needs to be refetched.
  } deriving (Eq, Generic, Show)

instance A.FromJSON Response

instance A.ToJSON Response where
  toEncoding = A.genericToEncoding A.defaultOptions


src/JMAP/Types/Error.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}

module JMAP.Types.Error
  ( ErrorResponse(..)
    -- * Smart Constructors
    -- $smartConstructors
  , unknownCapability
  , notJSON
  , notRequest
  , limit
  , errorResponse
  , ErrorType(..)
  ) where

import qualified Data.Aeson as A
import qualified Data.Aeson.Types as A (Parser)
import           Data.Text (Text)
import qualified Data.Text as T
import           GHC.Generics (Generic)
import           JMAP.Types.Base (Limit, URI)
import           Network.HTTP.RFC7807
import           Refined (unrefine)

-- | When an HTTP error response is returned to the client, the server
-- SHOULD return a JSON "problem details" object as the response body,
-- as per [RFC7807]. (§ 3.6.1)
newtype ErrorResponse
  = ErrorResponse (Rfc7807Error ErrorType () ())
  deriving (Eq, Generic, Show)

instance A.FromJSON ErrorResponse where
  parseJSON = A.withObject "ErrorResponse" $ \o ->
    fmap ErrorResponse $ Rfc7807Error
    <$> parseET o
    <*> o A..:? "title"
    <*> o A..:? "status"
    <*> o A..:? "detail"
    <*> o A..:? "instance"
    <*> o A..:? "error"
    <*> o A..:? "context"

instance A.ToJSON ErrorResponse where
  toJSON = A.Object . erPairs
  toEncoding = A.pairs . erPairs


-- | Helper: creates the key-value pairs used in the JSON
-- representation of 'ErrorResponse'.
erPairs :: forall kv. (A.KeyValue kv, Monoid kv) => ErrorResponse -> kv
erPairs (ErrorResponse a) = toKeyValue def a <> etPairs (type_ a)
  where def = defaultEncodingOptions

-- $smartConstructors
--
-- There is one smart constructor for each constructor of 'ErrorType'.

-- | Smart constructor for an 'ErrorResponse' with URN
-- "urn:ietf:params:jmap:error:unknownCapability".
unknownCapability :: URI -> ErrorResponse
unknownCapability u = errorResponse $ UnknownCapability u

-- | Smart constructor for an 'ErrorResponse' with URN
-- "urn:ietf:params:jmap:error:notJSON"
notJSON :: ErrorResponse
notJSON = errorResponse NotJSON

-- | Smart constructor for an 'ErrorResponse' with URN
-- "urn:ietf:params:jmap:error:notRequest"
notRequest :: ErrorResponse
notRequest = errorResponse NotRequest

-- | Smart constructor for an 'ErrorResponse' with URN
-- "urn:ietf:params:jmap:error:limit"
limit :: Limit -> ErrorResponse
limit l = errorResponse $ Limit l

-- | A general smart constructor for an 'ErrorResponse'; it fills in
-- all the fields that can be directly deduced from the 'ErrorType'
-- argument.
errorResponse :: ErrorType -> ErrorResponse
errorResponse a = ErrorResponse $ (rfc7807Error a)
  { title = Just $ errorTypeTitle a
  , detail = errorTypeDetail a
  }


-- | 'ErrorType' captures the minimum information necessary to convey
-- a particular JMAP error. The error types are defined in § 3.6.1.
data ErrorType
  = UnknownCapability !URI
    -- ^ The client included a capability in the "using" property of
    -- the request that the server does not support.
  | NotJSON
    -- ^ The content type of the request was not "application/json" or
    -- the request did not parse as I-JSON.
  | NotRequest
    -- ^ The request parsed as JSON but did not match the type
    -- signature of the Request object.
  | Limit !Limit
    -- ^ The request was not processed as it would have exceeded one
    -- of the request limits defined on the capability object, such as
    -- maxSizeRequest, maxCallsInRequest, or maxConcurrentRequests. A
    -- "limit" property MUST also be present on the "problem details"
    -- object, containing the name of the limit being applied.
  deriving (Eq, Generic, Show)

instance A.ToJSON ErrorType where
  toJSON = A.toJSON . errorTypeURN
  toEncoding = A.toEncoding . errorTypeURN


-- | Helper: creates the key-value pairs used in the JSON
-- representation of 'ErrorResponse' that are specific to a particular
-- 'ErrorType'.
etPairs :: forall kv. (A.KeyValue kv, Monoid kv) => ErrorType -> kv
etPairs (UnknownCapability u) = "unknownCapability" A..= u
etPairs (Limit l) = "limit" A..= l
etPairs _ = mempty

-- | Helper: parses an 'ErrorType' out of the same set of key-value
-- pairs that encode an 'ErrorResponse'.
parseET :: A.Object -> A.Parser ErrorType
parseET o = o A..: "type" >>= \case
  "urn:ietf:params:jmap:error:unknownCapability" ->
    UnknownCapability <$> o A..: "unknownCapability"
  "urn:ietf:params:jmap:error:notJSON" ->
    pure NotJSON
  "urn:ietf:params:jmap:error:notRequest" ->
    pure NotRequest
  "urn:ietf:params:jmap:error:limit" ->
    Limit <$> o A..: "limit"
  u -> fail $ "Did not recognize error type '"<>T.unpack u<>"'"

-- | Projects 'ErrorType' onto the "type" field of 'ErrorResponse'.
errorTypeURN :: ErrorType -> URI
errorTypeURN (UnknownCapability _) =
  "urn:ietf:params:jmap:error:unknownCapability"
errorTypeURN NotJSON =
  "urn:ietf:params:jmap:error:notJSON"
errorTypeURN NotRequest =
  "urn:ietf:params:jmap:error:notRequest"
errorTypeURN (Limit _) =
  "urn:ietf:params:jmap:error:limit"

-- | Projects 'ErrorType' onto the "title" field of 'ErrorResponse'.
errorTypeTitle :: ErrorType -> Text
errorTypeTitle (UnknownCapability _) =
  "Requested capability is not known to the server"
errorTypeTitle NotJSON =
  "Server was unable to parse request as I-JSON"
errorTypeTitle NotRequest =
  "Server did not recognize JSON document as a Request object"
errorTypeTitle (Limit _) =
  "Server did not process request as it would exceed limit"

-- | Projects 'ErrorType' onto the "detail" field of 'ErrorResponse'.
errorTypeDetail :: ErrorType -> Maybe Text
errorTypeDetail (UnknownCapability u) = Just $ "The client included the capability '"<>u<>"' in the 'using' property of the request. The server does not support that capability."
errorTypeDetail NotJSON = Just "The body of the request was either not a valid JSON document as defined in RFC-8259, or does not obey the conventions of the Internet JSON (I-JSON) format as defined in RFC-7493."
errorTypeDetail NotRequest = Just "The server recognized the body of the request as valid I-JSON, but didn’t recognize that JSON document as a Request object as defined in the JMAP spec (RFC-8620 § 3.3)."
errorTypeDetail (Limit l) = Just $ "The request was not processed as it would have exceeded the '"<>unrefine l<>"' limit of the capability 'urn:ietf:params:jmap:core'. Capabilities and their limits may be found in the Session object returned from the server in accordance with the JMAP spec (RFC-8620 § 2)."
